unlock-app/src/components/interface/members/airdrop/AirdropBulkForm.tsx

```diff
@@ -36,78 +39,103 @@ export function AirdropBulkForm({ lock, onConfirm }: Props) {
-              return
-            }
 
-            const member = {
+            return {
               ...record,
               recipient,
               manager,
+              balance,
+              line: line + 2,
             }
-            return member
           } catch (error) {
-            console.error(error)
+            linesWithError.push(line + 2) // index starts at 0, lines at 1 and we have the header row
+            console.error(`Failed to add `, item, error)
             return
           }
         })
       )
 
-      const filteredMembers = members.reduce<AirdropMember[]>(
-        (filtered, member) => {
+      const filteredMembers = members
+        .reduce<AirdropMember[]>((filtered, member) => {
           // filter null or undefined values
           if (!member) {
-            // keep track of discarded entries
-            discarded += 1
             return filtered
           }
 
-          // find existing member
-          const existingMember = filtered.find(
+          // find existing members
+          const existingMembers = filtered.filter(
             ({ recipient }) => recipient === member.recipient
           )
 
-          // if exist, discard the current entry to avoid duplicate.
-          if (existingMember) {
-            // keep track of discarded entries
-            discarded += 1
+          const existingBalance = member.balance || 0
+          const alreadyToBeAdded = existingMembers.reduce(
+            (total, existingMember) => total + existingMember.count,
+            0
+          )
+          const toBeAdded = member.count
+          if (
+            existingBalance + alreadyToBeAdded + toBeAdded >
+            (lock.maxKeysPerAddress || 1)
+          ) {
+            console.warn(`Discarded duplicate`, member)
+            duplicates.push(member)
+            discarded.push(member)
             return filtered
           }
 
           // push the item to array if new unique member found
           filtered.push(member)
-
           return filtered
-        },
-        []
-      )
+        }, [])
+        .slice(0, MAX_SIZE)
 
       // Notify how many loaded and discarded.
       ToastHelper.success(
-        `Loaded ${filteredMembers.length} members. ${discarded} members discarded.`
+        `Loaded ${filteredMembers.length} members. ${discarded.length} members discarded.`
       )
+      const errors: string[] = []
+      if (linesWithError.length > 0) {
+        errors.push(
+          `The following lines had errors in your CSV file and could not be imported: ${linesWithError.join(
+            ', '
+          )}.`
+        )
+      }
+      if (duplicates.length > 0) {
+        errors.push(
+          `The following recipients are duplicates and have been discarded: ${duplicates
+            .map((m) => `${m.recipient} (line ${m.line})`)
+            .join(', ')}`
+        )
+      }
+      if (errors.length > 0) {
+        setError(errors.join('\n\n'))
```

Comment:
`\n\n` wouldn't be displayed properly in the <p> tag below. 

Change it to be an array and handle formatting by looping using <p> or use <pre>